coreneuron/io/phase2.cpp

```diff
@@ -0,0 +1,1211 @@
+            } else if (s >= 0 && s < 1000) {  // ion
+                int etype = s;
+                /* if ion is SoA, must recalculate pdata values */
+                /* if ion is AoS, have to deal with offset */
+                Memb_list* eml = nt._ml_list[etype];
+                int edata0 = eml->data - nt._data;
+                int ecnt = eml->nodecount;
+                int esz = nrn_prop_param_size_[etype];
+                for (int iml = 0; iml < cnt; ++iml) {
+                    int* pd = pdata + nrn_i_layout(iml, cnt, i, szdp, layout);
+                    int ix = *pd;  // relative to the ion data
+                    nrn_assert((ix >= 0) && (ix < ecnt * esz));
+                    /* Original pd order assumed ecnt groups of esz */
+                    *pd = edata0 + nrn_param_layout(ix, etype, eml);
+                }
+            }
+        }
+    }
+
+    /* if desired, apply the node permutation. This involves permuting
+       at least the node parameter arrays for a, b, and area (and diam) and all
+       integer vector values that index into nodes. This could have been done
+       when originally filling the arrays with AoS ordered data, but can also
+       be done now, after the SoA transformation. The latter has the advantage
+       that the present order is consistent with all the layout values. Note
+       that after this portion of the permutation, a number of other node index
+       vectors will be read and will need to be permuted as well in subsequent
+       sections of this function.
+    */
+    if (interleave_permute_type) {
+        nt._permute = interleave_order(nt.id, nt.ncell, nt.end, nt._v_parent_index);
+    }
+    if (nt._permute) {
+        int* p = nt._permute;
+        permute_data(nt._actual_a, nt.end, p);
+        permute_data(nt._actual_b, nt.end, p);
+        permute_data(nt._actual_area, nt.end, p);
+        permute_data(nt._actual_v, nt.end,
+                     p);  // need if restore or finitialize does not initialize voltage
+        if (nt._actual_diam) {
+            permute_data(nt._actual_diam, nt.end, p);
+        }
+        // index values change as well as ordering
+        permute_ptr(nt._v_parent_index, nt.end, p);
+        node_permute(nt._v_parent_index, nt.end, p);
+
+#if DEBUG
+        for (int i=0; i < nt.end; ++i) {
+            printf("parent[%d] = %d\n", i, nt._v_parent_index[i]);
+        }
+#endif
+
+        // specify the ml->_permute and sort the nodeindices
+        for (auto tml = nt.tml; tml; tml = tml->next) {
+            if (tml->ml->nodeindices) {  // not artificial
+                permute_nodeindices(tml->ml, p);
+            }
+        }
+
+        // permute mechanism data, pdata (and values)
+        for (auto tml = nt.tml; tml; tml = tml->next) {
+            if (tml->ml->nodeindices) {  // not artificial
+                permute_ml(tml->ml, tml->index, nt);
+            }
+        }
```

Comment:
that can be done in one loop?

Reply:
depends of side effets of `permute_*` you know a lot of function here take `int*&` as parameters

Reply:
This is how SoA works you know, on index

Reply:
Done, let see what jenkins tell us, about that.